app/services/order_invoice_comparator.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class OrderInvoiceComparator
+  def can_generate_new_invoice?(current_state_invoice, latest_invoice)
+    # We'll use a recursive BFS algorithm to find if the invoice is outdated
+    # the root will be the order
+    # On each node, we'll a list of relevant attributes that will be used on the comparison
+    bfs(current_state_invoice.presenter, latest_invoice.presenter, invoice_generation_selector)
+  end
+
+  def can_update_latest_invoice?(current_state_invoice, latest_invoice)
+    bfs(current_state_invoice.presenter, latest_invoice.presenter, invoice_update_selector)
+  end
+
+  def bfs(node1, node2, attributes_selector)
+    simple_values1, presenters1 = attributes_selector.call(node1)
+    simple_values2, presenters2 = attributes_selector.call(node2)
+    return true if simple_values1 != simple_values2
+
+    return true if presenters1.size != presenters2.size
+
+    presenters1.zip(presenters2).each do |presenter1, presenter2|
+      return true if bfs(presenter1, presenter2,attributes_selector)
+    end
+    false
```

Comment:
Good idea with `zip`. In conjunction with my naming suggestion above, I would write this with `any?`:
```suggestion
    presenters1.zip(presenters2).any? do |presenter1, presenter2|
      different?(presenter1, presenter2, attributes_selector)
    end
```